File: backend/src/models/goal.js
const mongoose = require("mongoose");

const contributionSchema = mongoose.Schema({
  amount: {
    type: Number,
    required: [true, "Contribution must have an amount"],
  },
  date: {
    type: Date,
    default: Date.now, // Automatically set the contribution date
  },
  note: String,
});

const goalSchema = mongoose.Schema({
  coverImage: {
    type: String,
    required: [true, "Goals should have cover images"],
  },
  goalName: {
    type: String,
    required: [true, "Goals should have a name"],
  },
  goalAmount: {
    type: Number,
    required: [true, "Goals should have an amount"],
  },
  goalAmountContributed: {
    type: [contributionSchema], // Array of contributions
    default: [],
  },
  goalCurrency: {
    type: String,
    required: [true, "Goals should have currencies specified"],
  },
  goalDeadline: Date,
  goalNote: String,
  goalColour: {
    type: String,
    default: "PrimaryPurple",
  },
  userId: {
    type: String,
    required: [true, "Goals cannot be created without a user"],
  },
});

const Goal = mongoose.model("Goal", goalSchema);

module.exports = Goal;

File: Trezo/Trezo/Models/Goal.swift
//
//  Goal.swift
//  Trezo
//
//  Created by Bekithemba Mdluli on 2024/11/12.
//

import Foundation

// MARK: - Top-level response
struct GoalsResponse: Decodable {
    let status: String
    let results: Int
    let data: GoalsData
}

struct GoalsData: Decodable {
    let goals: [Goal]
}

// MARK: - Goal model
struct Goal: Decodable {
    let _id: String
    let coverImage: String
    let goalName: String
    let goalAmount: Double
    let goalAmountContributed: [Contribution]
    let goalCurrency: String
    let goalDeadline: Date?
    let goalNote: String?
    let goalColour: String
    let userId: String
    
    
    init(_id: String = "", coverImage: String = "", goalName: String = "", goalAmount: Double = 0.0, goalAmountContributed: [Contribution] = [], goalCurrency: String = "", goalDeadline: Date? = Date(), goalNote: String? = "", goalColour: String = "", userId: String = "") {
        self._id = _id
        self.coverImage = coverImage
        self.goalName = goalName
        self.goalAmount = goalAmount
        self.goalAmountContributed = goalAmountContributed
        self.goalCurrency = goalCurrency
        self.goalDeadline = goalDeadline
        self.goalNote = goalNote
        self.goalColour = goalColour
        self.userId = userId
    }
}
